ledger/nomos-ledger/src/mantle/leader.rs

```diff
@@ -0,0 +1,122 @@
+use crate::Balance;
+use cryptarchia_engine::Epoch;
+use mmr::MerkleMountainRange;
+use nomos_core::mantle::{
+    ops::leader_claim::{LeaderClaimOp, RewardsRoot, VoucherNullifier, VoucherCm},
+    Value,
+};
+use std::cmp::Ordering;
+
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LeaderState {
+    // current epoch
+    epoch: Epoch,
+    // vouchers that can be claimed in this epoch
+    // this is updated once at the start of each epoch from the root of
+    // the vouchers merkle tree
+    claimable_vouchers_root: RewardsRoot,
+    n_claimable_vouchers: u64,
+    // nullifiers of vouchers that have been claimed since genesis
+    nfs: rpds::HashTrieSetSync<VoucherNullifier>,
+    // rewards to be distributed
+    // at the start of each epoch this is increased by the amount of rewards
+    // that have been collected in the previous epoch.
+    // unclaimed rewards are carried over to the next epoch.
+    rewards: Value,
+    // Merkle tree of vouchers, vouchers can only be claimed with a delay
+    // of one epoch.
+    vouchers: MerkleMountainRange<VoucherCm, nomos_core::crypto::Hasher>,
+}
+
+#[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
+pub enum Error {
+    #[error("voucher nullifier already used")]
+    DuplicatedVoucherNullifier,
+    #[error("voucher not found")]
+    VoucherNotFound,
+    #[error("Cannot time travel to the past")]
+    InvalidEpoch { current: Epoch, incoming: Epoch },
+}
+
+impl LeaderState {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            epoch: 0.into(),
+            claimable_vouchers_root: [0; 32],
+            n_claimable_vouchers: 0,
+            nfs: rpds::HashTrieSetSync::new_sync(),
+            rewards: 0,
+            vouchers: MerkleMountainRange::new(),
+        }
+    }
+
+    #[must_use]
+    pub fn try_apply_header(self, epoch: Epoch, voucher_cm: VoucherCm) -> Result<Self, Error> {
+        Ok(self.update_epoch_state(epoch)?
+            .add_voucher(voucher_cm))
+    }
+
+    fn update_epoch_state(mut self, epoch: Epoch) -> Result<Self, Error> {
+        match epoch.cmp(&self.epoch) {
+            Ordering::Equal => Ok(self),
+            Ordering::Less => Err(Error::InvalidEpoch {
+                current: self.epoch,
+                incoming: epoch,
+            }),
+            Ordering::Greater => {
+                self.epoch = epoch;
+                self.claimable_vouchers_root = self.vouchers.frontier_root();
+                self.n_claimable_vouchers = self.vouchers.len() as u64;
+                // TODO: increase rewards, what about epoch jumps?
+                Ok(self)
+            }
+        }
+    }
+
+    fn add_voucher(self, voucher_cm: VoucherCm) -> Self {
+        Self {
+            vouchers: self.vouchers.push(voucher_cm),
+            ..self
+        }
+    }
```

Comment:
For testing purposes, can we increase `reward` by 1 for each voucher added? That way we can test the claim flow